src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,407 @@
+{
+    QStringList categories;
+    const QString name = params()["pluginName"].trimmed();
+
+    categories << SearchPluginManager::categoryFullName("all");
+    for (const QString &category : SearchPluginManager::instance()->getPluginCategories(name))
+        categories << SearchPluginManager::categoryFullName(category);
+
+    const QJsonArray result = QJsonArray::fromStringList(categories);
+    setResult(result);
+}
+
+void SearchController::pluginsAction()
+{
+    const QStringList allPlugins = SearchPluginManager::instance()->allPlugins();
+    setResult(getPluginsInfo(allPlugins));
+}
+
+void SearchController::installPluginAction()
+{
+    checkParams({"sources"});
+
+    const QStringList sources = params()["sources"].split('|');
+    for (const QString &source : sources)
+        SearchPluginManager::instance()->installPlugin(source);
+}
+
+void SearchController::uninstallPluginAction()
+{
+    checkParams({"names"});
+
+    const QStringList names = params()["names"].split('|');
+    for (const QString &name : names)
+        SearchPluginManager::instance()->uninstallPlugin(name.trimmed());
+}
+
+void SearchController::enablePluginAction()
+{
+    checkParams({"names", "enable"});
+
+    const QStringList names = params()["names"].split('|');
+    const bool enable = Utils::String::parseBool(params()["enable"].trimmed(), false);
+
+    for (const QString &name : names)
+        SearchPluginManager::instance()->enablePlugin(name.trimmed(), enable);
+}
+
+void SearchController::updatePluginsAction()
+{
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFinished, this, &SearchController::checkForUpdatesFinished);
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFailed, this, &SearchController::checkForUpdatesFailed);
+    pluginManager->checkForUpdates();
+}
+
+void SearchController::searchTimeout(const QString &id)
+{
+    ISession *session = sessionManager()->session();
+    auto searchHandlers = session->getData(SEARCH_HANDLERS).value<SearchHandlerDict>();
+
+    if (searchHandlers.contains(id)) {
+        LogMsg(tr("Deleting search session %1 from web session due to timeout").arg(id), Log::INFO);
+        SearchHandlerPtr searchHandler = searchHandlers[id];
+        searchHandler->cancelSearch();
+        searchHandlers.remove(id);
+        session->setData(SEARCH_HANDLERS, QVariant::fromValue(searchHandlers));
+
+        removeActiveSearch(session, id);
+    }
+}
+
+void SearchController::checkForUpdatesFinished(const QHash<QString, PluginVersion> &updateInfo)
+{
+    if (updateInfo.isEmpty()) {
+        LogMsg(tr("All plugins are already up to date."), Log::INFO);
+        return;
+    }
+
+    LogMsg(tr("Updating %1 plugins").arg(updateInfo.size()), Log::INFO);
+
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+    for (const QString &pluginName : updateInfo.keys()) {
```

Comment:
`for (const QString &pluginName : copyAsConst(updateInfo.keys())) {`

Reply:
`copyAsConst` seems to be required only when the method returns a copy of an existing container rather than creating it.